pkg/packetfilter/nftables/nftables.go

```diff
@@ -0,0 +1,511 @@
+	return p.insertRuleAtPosition(chain, rule, -1)
+}
+
+func (p *packetFilter) CreateIPHookChainIfNotExists(chain *packetfilter.ChainIPHook) error {
+	exists, err := p.chainExists(chain.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if chain exists")
+	}
+
+	if exists {
+		return nil
+	}
+
+	tx := p.nftables.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for submariner"),
+	})
+
+	chainPriority := iphookChainTypeToNftablesBasicPriority[chain.Type]
+	if chain.Priority == packetfilter.ChainPriorityFirst {
+		chainPriority += "-10"
+	}
+
+	nftChain := &knftables.Chain{
+		Name:     chain.Name,
+		Type:     ptr.To(iphookChainTypeToNftablesType[chain.Type]),
+		Hook:     ptr.To(iphookChainHookToNftablesHook[chain.Hook]),
+		Priority: ptr.To(chainPriority),
+	}
+	tx.Add(nftChain)
+	err = p.nftables.Run(context.TODO(), tx)
+
+	return errors.Wrap(err, "error creating IPHook Chain")
+}
+
+func (p *packetFilter) CreateChainIfNotExists(table packetfilter.TableType, chain *packetfilter.Chain) error {
+	exists, err := p.ChainExists(table, chain.Name)
+	if err != nil {
+		return errors.Wrap(err, "failed to check if chain exists")
+	}
+
+	if exists {
+		return nil
+	}
+
+	tx := p.nftables.NewTransaction()
+	tx.Add(&knftables.Table{
+		Comment: ptr.To("rules for submariner"),
+	})
+
+	tx.Add(&knftables.Chain{
+		Name: chain.Name,
+	})
+
+	err = p.nftables.Run(context.TODO(), tx)
+
+	return errors.Wrap(err, "error creating chain")
+}
+
+func (p *packetFilter) DeleteIPHookChain(chain *packetfilter.ChainIPHook) error {
+	return p.deleteChain(chain.Name)
+}
+
+func (p *packetFilter) DeleteChain(_ packetfilter.TableType, chain string) error {
+	return p.deleteChain(chain)
+}
+
+func (p *packetFilter) deleteChain(chain string) error {
+	tx := p.nftables.NewTransaction()
+	tx.Delete(&knftables.Chain{
+		Name: chain,
+	})
+
+	err := p.nftables.Run(context.TODO(), tx)
+	if knftables.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "error deleting chain")
+}
+
+func (p *packetFilter) ClearChain(_ packetfilter.TableType, chain string) error {
+	tx := p.nftables.NewTransaction()
+	tx.Flush(&knftables.Chain{
+		Name: chain,
+	})
+
+	err := p.nftables.Run(context.TODO(), tx)
+	if knftables.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "error clearing chain")
+}
+
+func (p *packetFilter) Delete(_ packetfilter.TableType, chain string, rule *packetfilter.Rule) error {
+	tx := p.nftables.NewTransaction()
+	tx.Delete(&knftables.Rule{
+		Chain: chain,
+		Rule:  ToRuleSpec(rule).String(),
+	})
+
+	err := p.nftables.Run(context.TODO(), tx)
+	if knftables.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrap(err, "error deleting rule")
+}
+
+func (p *packetFilter) List(_ packetfilter.TableType, chain string) ([]*packetfilter.Rule, error) {
+	existingRules, err := p.nftables.ListRules(context.TODO(), chain)
```

Comment:
The docs for `ListRules`  says "_the Rule objects will have their `Comment` and `Handle` fields filled in, but *not* the actual `Rule` field._". I also verified this in the code. I think we need to also set the `Comment` field to the rule spec string on insert/append.